src/client/dfuse/il/int_posix.c

```diff
@@ -377,8 +378,10 @@ ioil_fini(void)
 			ioil_shrink_pool(pool);
 	}
 
-	if (ioil_iog.iog_daos_init)
+	if (ioil_iog.iog_daos_init) {
+		daos_eq_destroy(ioil_eqh, 0);
```

Comment:
This may be problematic. The destructor function is only executed by the main thread. The event queue created by other child threads are leaked. We need to maintain a list of "ioil_eqh" allocated by all threads and destroy all of them here.

Reply:
yea im not sure how to handle this. i tried doing atexit handlers, but those don't get executed in case of fork().
in case of threads, yea having a list of eqs might work.

Reply:
For fork(), we could intercept fork(). We can tell we are in parent or child process from the pid returned by fork(). We can zero ioil_eqh in child process. It will be re-created when necessary. destructor function will be called by both parent and child processes.

Reply:
hmm i think the destructor function is only called on the parent as i can tell.

Reply:
I was not sure either, so I wrote a mini code to verify. It was confirmed that both parent and child process call the destructor function.

Example code,

#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>

int tid;

static __attribute__((destructor)) void fini()
{
        printf("In fini(). tid = %d\n", tid);
}

void forkexample()
{
        tid = fork();
        // child process because return value zero
        if (tid == 0)
                printf("Hello from Child!\n");

        // parent process because return value non-zero.
        else
                printf("Hello from Parent!\n");
}

int main()
{
        forkexample();
        return 0;
}

Reply:
try to call exec in the child

Reply:
i have added a print in ioil_fini() and run fio with 32 processes, and i only see one call.
i assume because of this:
"Upon a successful call to one of the [exec(3)](https://man7.org/linux/man-pages/man3/exec.3.html) functions, all registrations are removed."

Reply:
yes. calling exec() would be different. It starts a fresh process. "ioil_eqh" created by the parent still will be destroyed.

Reply:
@mchaarawi I checked fio code again. With your configure file, fio call pthread_create() to spawn new threads. It does not call fork/exec. ioil_fini() is expected to be called only one time since there is only one process in fio.

Reply:
im using the non-thread option in fio which does use fork. let's discuss offline

Reply:
Have a look at L349-351 above, there are lots of cases where this destructor is being called at least from different threads, I'm not sure if it's called from multiple threads in the same process however.  Without the thread check we were seeing segfaults however.

Reply:
yes i saw those. in my testing, the destructor is called only once. i verified this with Lei and it's because fio calls exit on forked processes which does not call this destructor.

Reply:
Right. We need to intercept "_exit()" (not "exit()") and call daos_eq_destroy() for all allocated eqs there too. Destructor function is still executed for processes calling "exit()" in my test.